File: 백준/Silver/15650. N과 M （2）/N과 M （2）.js
const [N, M] = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split(" ")
  .map(Number);

const sequence = [];

const dfs = (count, start) => {
  if (count === M) {
    console.log(sequence.join(" "));
    return;
  }

  for (let i = start; i < N; i++) {
    if (!sequence.includes(i + 1)) {
      sequence.push(i + 1);
      start++;
      dfs(count + 1, start);
      sequence.pop();
    }
  }
};

dfs(0, 0);

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py

N, M = map(int, input().split())

# 1부터 N까지 자연수에서 중복없이 M개 고른 수열


stack = []


def dfs(depth):
    if depth == M:
        print(*stack)

    for i in range(1, N+1):
        if i in stack or (stack and stack[-1] > i):
            continue
        stack.append(i)
        dfs(depth+1)
        stack.pop()


dfs(0)